proxy/src/control/destination/mod.rs

```diff
@@ -227,30 +194,15 @@ where
             let update = try_ready!(up).expect("destination stream must be infinite");
 
             match update {
-                Update::Insert(addr, meta) => {
-                    // Construct a watch for the `Labeled` middleware that will
-                    // wrap the bound service, and insert the store into our map
-                    // so it can be updated later.
-                    let (labels_watch, labels_store) = Watch::new(meta.metric_labels);
-                    self.metric_labels.insert(addr, labels_store);
-
-                    let endpoint = Endpoint::new(addr, labels_watch.clone());
+                Update::Insert(addr, meta) | Update::ChangeMetadata(addr, meta) => {
+                    let endpoint = Endpoint::new(addr, meta.metric_labels.clone());
 
                     let service = self.bind.bind(&endpoint).map_err(|_| ())?;
 
                     return Ok(Async::Ready(Change::Insert(addr, service)));
```

Comment:
`Change::Insert` should (eventually) be renamed `Change::Upsert` or similar. Let's add a comment here explaining that we expect `Change::Insert` to be processed with "upsert" semantics, and let's add a comment to `Change::Insert`/`Change::Upsert` indicating that when a `Change::Insert`/`Change::Upsert` is received, one must replace and rebind any equivalent extant service with the one received from the event.

It seems like in the future we could probably combine `Change::Insert` and `Change::ChangeMetadata` in a similar way? If so, let's do that in a follow-up.

Reply:
I don't think we should rely on jargon like Upsert